marlowe/src/Language/Marlowe/Scripts.hs

```diff
@@ -171,6 +179,122 @@ mkValidator :: MarloweParams -> Scripts.ValidatorType MarloweStateMachine
+    let slotConfig = def :: TimeSlot.SlotConfig
+    let ownInput = case findOwnInput ctx of
+            Just i -> i
+            _      -> traceError "I0" {-"Can't find validation input"-}
+    let scriptInValue = txOutValue $ txInInfoResolved ownInput
+    let (minTime, maxTime) =
+            case txInfoValidRange scriptContextTxInfo of
+                Interval.Interval (Interval.LowerBound (Interval.Finite l) True) (Interval.UpperBound (Interval.Finite h) False) -> (l, h)
+                _ -> traceError "R0"
+    let timeToSlot = TimeSlot.posixTimeToEnclosingSlot slotConfig
+    let minSlot = timeToSlot minTime
+    let maxSlot = timeToSlot maxTime
+    let interval = (minSlot, maxSlot)
+    let positiveBalances = traceIfFalse "B0" $ validateBalances marloweState
+
+    {-  We do not check that a transaction contains exact input payments.
+        We only require an evidence from a party, e.g. a signature for PubKey party,
+        or a spend of a 'party role' token.
+        This gives huge flexibility by allowing parties to provide multiple
+        inputs (either other contracts or P2PKH).
+        Then, we check scriptOutput to be correct.
+     -}
+    let inputsOk = validateInputs inputs
+
+    -- total balance of all accounts in State
+    -- accounts must be positive, and we checked it above
+    let inputBalance = totalBalance (accounts marloweState)
+
+    -- ensure that a contract TxOut has what it suppose to have
+    let balancesOk = traceIfFalse "B1" $ inputBalance == scriptInValue
+
+    let preconditionsOk = positiveBalances && balancesOk
+
+    let txInput = TransactionInput {
+            txInterval = interval,
+            txInputs = inputs }
+
+    let computedResult = computeTransaction txInput marloweState marloweContract
+    -- let computedResult = TransactionOutput [] [] (emptyState minSlot) Close
+    case computedResult of
+        TransactionOutput {txOutPayments, txOutState, txOutContract} -> do
+            let marloweData = MarloweData {
+                    marloweContract = txOutContract,
+                    marloweState = txOutState }
+
+                payoutsByParty = AssocMap.toList $ foldMap payoutByParty txOutPayments
+                payoutsOk = payoutConstraints payoutsByParty
+                checkContinuation = case txOutContract of
+                    Close -> True
+                    _ -> let
+                        totalIncome = foldMap collectDeposits inputs
+                        totalPayouts = foldMap snd payoutsByParty
+                        finalBalance = inputBalance + totalIncome - totalPayouts
+                        outConstrs = OutputConstraint
+                                    { ocDatum = marloweData
+                                    , ocValue = finalBalance
+                                    }
+                        in checkOwnOutputConstraint ctx outConstrs
```

Comment:
FYI: [`checkOwnOutputConstraints`](https://github.com/input-output-hk/plutus-apps/blob/a6868cde87600051a4cc808777d665d0b859853c/plutus-ledger/src/Ledger/Constraints/OnChain.hs#L42-L51) contains logic related to min-ADA. At least one output must simultaneously satisfy these conditions:
1. The datum hash matches.
2. The non-ADA value exactly matches the constraint.
3. The ADA value is `>=` the constraint and `<=` 2 ADA plus the constraint.

This shouldn't adversely affect this validator.